### PR TITLE
Neural network classifier benchmark

### LINKED ISSUE #12: NeuralNetworkClassifier benchmark.
This benchmark might test:

- OpflowQNN / CircuitQNN
    - qasm / statevector

and more.

### PR DESCRIPTION
**Pull request by dlasecki**
### Summary

Resolves #12.

### Details and comments

Implemented 3 benchmarks that use OpflowQNN, CircuitQNN and VQC as classifiers. Benchmark based on the following tutorial: https://qiskit.org/documentation/machine-learning/tutorials/02_neural_network_classifier_and_regressor.html.

### CODE REVIEW

**Review by manoelmarques — DISMISSED** [review]

**Comment by dlasecki** [comment]
> Thanks for kicking off benchmarks in QML. Looks good and can be merged, I think.
> Let me know if these ideas are better to be implemented in a separate PR:
> 
> * I'd rather fix the datasets in the benchmarks, no randomization at all.

Currently, I use a fixed seed for both data generation and for algorithms (in case they use some randomness  [probably not but I did not check]). As a result, datasets are virtually fixed. However, there might be some benefit of making them explicit and verifying that they do not represent an unfortunate configuration for the algorithm. 

> * Make the datasets larger, not only 2 inputs and 20 samples. I think these tests are fast, please correct me if I'm wrong.

@manoelmarques checked that the "quick" run of all benchmarks in this PR (which means that asv does not repeat them for more accurate time measurement) lasts about 4 hours which is very long. I had the same experience when running them locally and it was quite troubling. It might actually be too long... Timed methods do not exceed 4 minutes for fitting and are almost instantaneous for predicting and scoring. I am now trying to move some initializations from setup to init methods such that they will not be repeated every time. Let's see how much it helps.

> * Have different datasets.

Sure, I intended to leave room for the mentorship project in this regard.

> * More emphasize on multiclass classification (definitely a separate task).

Yes, and also might be a good task for the mentorship project.

> * Score itself can be a measurement once we fix the datasets. In general, score should only go up.

We are not interested in evaluating the model here. Just execution time.

> * Score on an unseen test dataset, not on the labels that are used to fit a model. If a score on the train/validation dataset is going up, but going down on a test dataset, then we overfit.

Yes, but we are not interested in score here, just execution time. Do you think that fitting on the same dataset would somehow make it faster (i.e. not representative) when compared to a different dataset?

> * May be memory measurements is also a good idea.

asv can do that when we create methods with a `mem_` prefix. I would also leave it for a separate task/your mentee.

**Comment by adekusar-drl** [comment]
> 
> 
> > Thanks for kicking off benchmarks in QML. Looks good and can be merged, I think.
> > Let me know if these ideas are better to be implemented in a separate PR:
> > 
> > * I'd rather fix the datasets in the benchmarks, no randomization at all.
> 
> Currently, I use a fixed seed for both data generation and for algorithms (in case they use some randomness [probably not but I did not check]). As a result, datasets are virtually fixed. However, there might be some benefit of making them explicit and verifying that they do not represent an unfortunate configuration for the algorithm.

Yeah, this is clear. But having them fixed completely has more benefits, e.g. you copy/paste it and you are sure you obtain the same results.

> 
> > * Make the datasets larger, not only 2 inputs and 20 samples. I think these tests are fast, please correct me if I'm wrong.
> 
> @manoelmarques checked that the "quick" run of all benchmarks in this PR (which means that asv does not repeat them for more accurate time measurement) lasts about 4 hours which is very long. I had the same experience when running them locally and it was quite troubling. It might actually be too long... Timed methods do not exceed 4 minutes for fitting and are almost instantaneous for predicting and scoring. I am now trying to move some initializations from setup to init methods such that they will not be repeated every time. Let's see how much it helps.

Oh, 4 hours, I was not expecting such a long duration. Then if we want to test, say, on a slightly large dataset, e.g. 100 samples(and this is still a pretty small dataset) how do we tackle the problem?

> > * Score on an unseen test dataset, not on the labels that are used to fit a model. If a score on the train/validation dataset is going up, but going down on a test dataset, then we overfit.
> 
> Yes, but we are not interested in score here, just execution time. Do you think that fitting on the same dataset would somehow make it faster (i.e. not representative) when compared to a different dataset?
> 

The idea is not about speed, but quality of the model and how quality evolves across multiple runs of the test. So, we fix a dataset (no randomization), score the model and observe how the score evolves in time. We can have such a benchmark in a separate test, of course. I hope, it makes sense and implementable in asv.

**Review by manoelmarques — APPROVED** [review]

**Review by adekusar-drl — APPROVED** [review]
Looks good, thanks a lot! We'll continue in the Qiskit Advocate program.

**Comment by adekusar-drl** [comment]
> 
> I improved the performance today. Now, all benchmarks run in about 40-60 minutes in the quick mode which is more reasonable.
> 

So, how did you achieve that?

@manoelmarques I think we can have an arbitrary measurement? In Terra I implemented a couple of tests where we measure circuit depth.

**Comment by dlasecki** [comment]
> > I improved the performance today. Now, all benchmarks run in about 40-60 minutes in the quick mode which is more reasonable.
> 
> So, how did you achieve that?
> 
> @manoelmarques I think we can have an arbitrary measurement? In Terra I implemented a couple of tests where we measure circuit depth.

I moved scoring and predicting to a separate benchmark with fixed optimizers and loss functions (they only play an important role for fitting). Note that asv does not guarantee that timed methods are executed in order of their declaration and a model has to be fitted before scoring and predicting. For that reason, I needed to have a separate instance of a fitted model in the original version used for scoring and predicting exclusively. It was being created for all combinations of parameters - backends, optimizers and loss functions. It took time and was not really necessary. In the improved version scoring and predicting are only tested on 2 backends; the rest is fixed. Another option, I suppose, would have been a deep copy of a once fitted classifier and keeping things in the same benchmark class but it would still have introduced an unnecessary overhead.
